kafkakewl-metrics/src/main/scala/com/mwam/kafkakewl/metrics/services/HighOffsetMetricsExposer.scala

```diff
@@ -8,80 +8,117 @@ package com.mwam.kafkakewl.metrics.services
+final case class MetricUpdate(metric: Metric.Counter[Long], increment: Long) {
+  def applyChanges(): UIO[Unit] = {
+    metric.incrementBy(increment)
+  }
+}
+
+final case class PartitionMetrics(partitionMetrics: Map[Int, (Option[Metric.Counter[Long]], Long)]) {
+  def updateWithNewOffset(
+      newOffsets: Map[KafkaTopicPartition, KafkaTopicPartitionInfo],
+      config: TopicsConfig
+  ): (PartitionMetrics, List[MetricUpdate], Long) = {
+    var totalChange = 0L
+    var metricUpdates = List.empty[MetricUpdate]
+    var newPartitionMap = partitionMetrics
+
+    for ((topicPartition, topicPartitionInfo) <- newOffsets) {
+      newPartitionMap = newPartitionMap.updatedWith(topicPartition.partition) {
+        case Some((counter, lastValue)) =>
+          val increase = topicPartitionInfo.endOffset - lastValue
+          totalChange = totalChange + increase
+          if (config.perPartitionHighOffsetMetrics) {
+            metricUpdates = MetricUpdate(counter.get, increase) +: metricUpdates
+          }
+          Some((counter, topicPartitionInfo.endOffset))
+        case None =>
+          totalChange = totalChange + topicPartitionInfo.endOffset
+
+          val counter = if (config.perPartitionHighOffsetMetrics) {
+            val counter = config.newPartitionHighOffsetMetric(topicPartition.topic, topicPartition.partition)
+            metricUpdates = MetricUpdate(counter, topicPartitionInfo.endOffset) +: metricUpdates
+            Some(counter)
+          } else {
+            None
+          }
+          Some((counter, topicPartitionInfo.endOffset))
+      }
+    }
+
+    (PartitionMetrics(newPartitionMap), metricUpdates, totalChange)
+  }
+
+  def removePartitions(topicPartitions: Set[KafkaTopicPartition], config: TopicsConfig): (PartitionMetrics, List[MetricUpdate], Long) = {
+    var totalChange = 0L
+    var metricUpdates = List.empty[MetricUpdate]
+    var newPartitionMap = partitionMetrics
+
+    for (topicPartition <- topicPartitions) {
+      newPartitionMap = newPartitionMap.get(topicPartition.partition) match {
+        case Some((counter, lastValue)) =>
+          totalChange = totalChange - lastValue
+          if (config.perPartitionHighOffsetMetrics) {
+            metricUpdates = MetricUpdate(counter.get, -lastValue) +: metricUpdates
+          }
+          newPartitionMap.removed(topicPartition.partition)
+
+        case None =>
+          newPartitionMap
+
+      }
+    }
+
+    (PartitionMetrics(newPartitionMap), metricUpdates, totalChange)
+  }
+
+  def nonEmpty: Boolean = partitionMetrics.nonEmpty
+}
+
+object PartitionMetrics {
+  def empty: PartitionMetrics = PartitionMetrics(Map.empty[Int, (Option[Metric.Counter[Long]], Long)])
 }
 
-case class MetricMap(metricMap: Map[String, (Map[Int, (Metric.Counter[Long], Long)], Metric.Counter[Long])]) {
+final case class MetricMap(metricMap: Map[String, (PartitionMetrics, Metric.Counter[Long])]) {
```

Comment:
How about renaming this to `TopicsMetricsState` or sg along those lines? (The reason of the `State` suffix is that perhaps we'll need the `TopicsMetrics` / `TopicMetrics` names for domain types, holding all topic metrics including consumer statuses etc... At least in legacy it's like that. These types here aren't really domain types, more like an implementation helper types to expose the topics' metrics)

The tuple in the map could have its own type:

```scala
final case class TopicMetricsState(partitionMetrics: PartitionMetricsState, aggregateHighOffsetCounter: Metric.Counter[Long])
```
 so that it's clearer which part does what.

This means renaming `PartitionMetrics` to `TopicPartitionsMetricsState` and it also can have a new type for the map's value:

 ```scala
final case class TopicPartitionMetricsState(highOffsetCounter: Option[Metric.Counter[Long]], highOffset: Long)
```